MSPI_DW DMA and Async support

Hi-Im-David opened the pull request:
PR for the mspi_dw driver to support Async and DMA transactions and support for the new nrf-mspi peripheral (name tbc). This peripheral is similar to the exmif in nRF54H20, using the same core designware IP, but has been configured to support DMA and slave mode. The DMA being used is part of the wrapper and not the core IP.

Concurrent Aync transactions and XiP support for nrf-mspi will be added in a future PR. The DMA support uses the DMA in a vendor specific wrapper.

Hi-Im-David committed:
drivers: pinctrl: nrf: add support for MSPI

Support for new MSPI peripheral where there is no PSEL
so pins are setup through CTRLSEL.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David committed:
dts: mspi: Align op-mode binding with mspi.h enum

enum mspi_op_mode in mspi.h has different syntax to this binding.
Aligning these will allow for cleaner code in the implmented drivers.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

swift-tk reviewed (commented):
Please add some explanation/doc on how should `MSPI_BUS_XFER_TIMEOUT` be used.

Hi-Im-David commented:
> Please add some explanation/doc on how should `MSPI_BUS_XFER_TIMEOUT` be used.

@swift-tk the idea was that the user could setup separate callback functions for when the transfer completes and when the transfer times out. I'm now realising this probably isn't needed as you can just use MSPI_BUS_XFER_COMPLETE with the error code ETIMEDOUT?

swift-tk commented:
> > Please add some explanation/doc on how should `MSPI_BUS_XFER_TIMEOUT` be used.
> 
> @swift-tk the idea was that the user could setup separate callback functions for when the transfer completes and when the transfer times out. I'm now realising this probably isn't needed as you can just use MSPI_BUS_XFER_COMPLETE with the error code ETIMEDOUT?

I'm not against adding a new event and callback, just that it needs to show a meaningful use case. But I think it is clearier as TIMEOUT may not necessarily be related to a (DMA)transfer directly.

BTW, you need to resolve conflict from main.

Hi-Im-David committed:
drivers: mspi: Add timeout callback to MSPI API

This is a new callback option that drivers can use for when a request
or xfer has timed out. E.g if an Async RX request never received
anything, this callback can be used so a user can clean up their
application.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1193,6 +1642,43 @@ static int api_transceive(const struct device *dev,
+		return -EINVAL;
+	}
+
+	if (evt_type != MSPI_BUS_XFER_COMPLETE) {
```
Good spot :)

Hi-Im-David commented on drivers/mspi/mspi_dw.h:
```diff
@@ -245,3 +262,32 @@ static void reg_write(uint32_t data, const struct device *dev, uint32_t off)
 	}
 #endif
+
+struct mspi_dw_config {
```
Nope, this was due a previous order I had the #include of this file needing it. Thanks

swift-tk reviewed (commented):
Please also update the mspi doc for the added kconfig option and timeout callback.
/doc/hardware/peripherals/mspi.rst
Explain a little bit of how the timeout callback should be used there.

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -283,62 +529,93 @@ static bool read_rx_fifo(const struct device *dev,
+		if (xfer.async) {
+			/* Cancel the timeout timer since transaction completed normally */
+			k_timer_stop(&dev_data->async_timeout_timer);
+			write_ssienr(dev, 0);
```
Why is the controller disabled here? If XIP is enabled, this can lead to a crush.

anangl reviewed (changes requested):
I haven't managed to go through all the code yet but there is already quite a few issues to be addressed.

Hi-Im-David commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +97,234 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+#define MAX_NUM_JOBS 4
+/* Just support 1 trasaction for each peripheral as concurrent transactions aren't supported yet*/
+#define MAX_CONCURR_TRANSACTIONS 1
+#define NUM_LISTS DT_NUM_INST_STATUS_OKAY(MSPI_DT_DRV_COMPAT)
+#define DMA_TRANSFER_LIST_SIZE (sizeof(QSPI_TRANSFER_LIST_Type) + sizeof(EVDMA_JOB_Type) * \
+				MAX_NUM_JOBS * MAX_CONCURR_TRANSACTIONS)
+#define DMA_TRANSFER_LIST_ALIGN 4
+K_MEM_SLAB_DEFINE(dma_transfer_list_slab, DMA_TRANSFER_LIST_SIZE, NUM_LISTS,
+		  DMA_TRANSFER_LIST_ALIGN);
```
I couldn't find a way to do it without mem slabs. The problem I see is the pointer that stores the transfer list would need to be allocated enough memory. The amount of memory to allocate will be vendor specific so the amount of memory that is being used needs to be defined by `mspi_dw_vendor_specific.h` but the `mspi_dw_data` struct needs to be compiled before the `mspi_dw_vendor_specific.h` file since the struct is used in the file. 

If i set a void pointer, that lives in `mspi_dw_data`, to a transfer list created in `vendor_specific_setup_dma_xfer` without `k_mem_slab_alloc`, that transfer list will live on the stack and not be safe once we exit the function.

Do you have any advice on how else I would get around this?

Hi-Im-David commented on drivers/mspi/mspi_dw.c:
```diff
@@ -814,10 +1091,157 @@ static void tx_control_field(const struct device *dev,
+}
+#endif /* CONFIG_MSPI_DMA */
+
+static int start_next_packet_pio(const struct device *dev, k_timeout_t timeout)
```
I've brought both back into common `start_next_packet`, let me know if this is sufficient. I've tested both using PIO and DMA and seems to be working fine. Originally I had separated this to make rebasing easier but never moved back.

anangl commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +97,228 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+	preg->CONFIG.TXBURSTLENGTH = (config->tx_fifo_depth_minus_1+1)-config->dma_tx_data_level;
+	preg->CONFIG.RXBURSTLENGTH = config->dma_rx_data_level+1;
+	preg->DMA.CONFIG.LISTPTR = (uint32_t)transfer_list;
+	preg->INTEN = BIT(QSPI_INTEN_DMADONE_Pos);
```
This clears `BIT(QSPI_INTENSET_CORE_Pos)`, why?

Hi-Im-David replied:
setting QSPI_INTENSET_CORE_Pos here seems to brick the system, not sure why but left a comment to address in future as I'm assuming this is needed for XiP? The system will still work in FIFO mode since this DMA function will never be called.

Hi-Im-David commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +97,228 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+/* Number of jobs needed for transmit trasaction */
+#define MAX_NUM_JOBS 4
+/* Just support 1 trasaction for each peripheral as concurrent transactions aren't supported yet*/
+#define MAX_CONCURR_TRANSACTIONS 1
```
By this, I mean multiple requests of mspi_transceive() that get buffered by the driver. Not sure if this was going to be needed but currently if another request is made while a transaction is already in progress it will just reject the request and say "Async transfer already in progress".

Hi-Im-David commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1261,39 +1581,67 @@ static int api_transceive(const struct device *dev,
+					k_timer_stop(&dev_data->async_timeout_timer);
+					cleanup_async_on_error(dev);
+					dev_data->async_in_progress = false;
+				}
```
I have kept `async_in_progress` flag to allow the driver to still return -EBUSY but I have cleaned up this logic by adding more common code to `cleanup_async_on_error()` and  adjusting how `_api_transceive()` and `api_transceive()` handle the ctx_lock, pm and sync/async return errors.

Hi-Im-David committed:
drivers: mspi_dw: Add support for asynchronous transfers

Handling of asynchronous transfers uses the system workqueue,
hence they are not available when multithreading is disabled.

Also add missing dependency on multithreading in the
MSPI_DW_HANDLE_FIFOS_IN_SYSTEM_WORKQUEUE Kconfig option.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

Hi-Im-David committed:
drivers: mspi_dw: Add support for slave mode

MSPI slave mode is selected through devicetree using the
op-mode property. Mode selected by SSIISMST bit in the
CTRLR0 register. EXMIF can only be Master (controller).

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David committed:
drivers: mspi: mspi_dw: Add DMA support

Initial DMA support. DMA supports implementation of SSI IP but using vendor
specific DMA in the wrapper. The setup of the DMA is done in
mspi_dw_vendor_specific.h.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David committed:
dts: nordic: Add nrf-qspi-v2 binding

The nrf-qspi-v2 peripheral is similar to EXMIF on nrf54h20
but supports DMA and slave-mode. The wrapper around
the SSI IP is also different with DMA features.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +97,223 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+#define MSPI_DW_DMA_XFER_LIST_GET(inst) &mspi_dw_##inst##_transfer_list
+#define MSPI_DW_DMA_JOBLIST_GET(inst) &mspi_dw_##inst##_joblist
+
+static inline void vendor_specific_start_dma_xfer(const struct device *dev)
```
I have resolved an issue which means we do not need seperate `vendor_specific_start_dma_xfer()` and `vendor_specific_setup_dma_xfer()` functions but I still believe we need a seperate `vendor_specific_dma_accessible_check()` function even if we are now calling it from just `start_next_packet` now. I think this check should happen before any configuration of the peripheral so it doesn't remain initialised on error. I guess we could also call the finalize_transceive/finalize_packet functions to cleanup on the error but not sure this is a better solution?

anangl commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +101,217 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+	preg->DMA.CONFIG.LISTPTR = (uint32_t)transfer_list;
+
+	preg->INTENSET = BIT(QSPI_INTENSET_CORE_Pos)
+		       | BIT(QSPI_INTENSET_DMADONE_Pos);
```
This is set in `vendor_specific_init()`. Why repeat it here?

anangl commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +101,217 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+
+/* Empty macros for generic case - no vendor-specific data */
+#define VENDOR_SPECIFIC_DATA_DEFINE(inst)
+#define VENDOR_SPECIFIC_DATA_GET(inst) NULL
```
Instead of repeating the empty definitions for every case where vendor specific data is not used, we could have at the end of this file:
```
#ifndef VENDOR_SPECIFIC_DATA_DEFINE
#define VENDOR_SPECIFIC_DATA_DEFINE(inst)
...
```

anangl commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +101,217 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+			transfer_list->tx_job = &joblist[0];
+		} else {
+			transfer_list->tx_job = &joblist[job_idx];
+		}
```
tx_job cannot be empty, see: https://github.com/Hi-Im-David/zephyr/blob/6875795cd4d34e8831873972cbc2c286dcd24b6b/drivers/mspi/mspi_dw.c#L1097-L1101

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -162,18 +173,18 @@ static void call_user_callback_with_context(const struct device *dev,
 
 	if (!(packet->cb_mask & BIT(evt_type)) ||
 	    !dev_data->cbs[evt_type]) {
+		LOG_DBG("Callback type not setup");
```
Is this really valuable to always log "Calling user function ..." and additionally the above if there's no callback? If so, these changes should be moved to the commit that introduces `call_user_callback_with_context()`.

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -112,6 +113,11 @@ struct mspi_dw_config {
+	uint8_t dma_tx_data_level;
+	uint8_t dma_rx_data_level;
+#endif
+	void const *vendor_specific_data;
```
This pointer should stay `void *` as the pointed object does not necessarily must be read-only in all cases. Instead, use `(void *)` casting in the `VENDOR_SPECIFIC_DATA_GET()` definition for `nordic_nrf_qspi_v2`.

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1211,95 +1251,123 @@ static int start_next_packet(const struct device *dev)
-		tx_data(dev, packet);
+		/* Enable interrupts now and wait until the packet is done unless async. */
+		write_imr(dev, imr);
+
```
nit: unneeded empty line

anangl commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +97,226 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+			joblist[1] = EVDMA_NULL_JOB();
+			transfer_list->tx_job = &joblist[1];
+			transfer_list->rx_job = &joblist[0];
+
```
ditto

anangl commented on drivers/mspi/mspi_dw_vendor_specific.h:
```diff
@@ -97,7 +97,226 @@ static inline int vendor_specific_xip_disable(const struct device *dev,
+			joblist[job_idx]   = EVDMA_NULL_JOB();
+
+			transfer_list->tx_job = &joblist[0];
+			transfer_list->rx_job = &joblist[job_idx-1];
```
I think it would be clearer to have:
```
			transfer_list->rx_job = &joblist[job_idx];
			joblist[job_idx++] = EVDMA_JOB(packet->data_buf, packet->num_bytes,
						       EVDMA_ATTRIBUTE);
```
Similarly for TX jobs - `transfer_list->tx_job = &joblist[job_idx];` could be put at the very beginning of the common path.

Hi-Im-David commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1211,95 +1251,123 @@ static int start_next_packet(const struct device *dev)
-		uint8_t start_level = dev_data->dummy_bytes != 0
-				    ? dev_config->max_queued_dummy_bytes - 1
-				    : dev_config->tx_fifo_depth_minus_1;
+	if (dev_data->xfer.xfer_mode == MSPI_PIO) {
```
Compliance checks fail with the error `BRACES: braces {} must be used on all arms of this statement due to the 
```
	} else
#endif
	{
```

I could do something like:
```
#if defined(CONFIG_MSPI_DMA)
	if (dev_data->xfer.xfer_mode == MSPI_DMA) {
		...
	} else {
#endif
	...
#if defined(CONFIG_MSPI_DMA)
}
#endif
```
But this is a bit messy? Or would you prefer this as it avoids the extra if statement when using PIO mode

anangl replied:
Or you could use for example:
```
	if (0) {
#if defined(CONFIG_MSPI_DMA)
	} else if (dev_data->xfer.xfer_mode == MSPI_DMA) {
		...
#endif
	} else {
```
which is also not very nice, but yes, I think it is worth using even something that weird to keep the PIO path the default one, as it was before this PR.
Actually, maybe what you proposed is not that bad. The PIO code would then not need to be indented, what would shorten the diff significantly.

carlescufi commented:
@swift-tk could you please take another look?
